includes/Modules/Analytics/Measurement_Events/Measurement_Event.php

```diff
@@ -0,0 +1,174 @@
+ * Use to instantiate MeasurementEvent objects
+ *
+ * @class Measurement_Event_Builder
+ */
+class Measurement_Event_Builder {
+	/**
+	 * Associative array of event configuration containing event attributes
+	 *
+	 * @var array
+	 */
+	private $configuration;
+
+	/**
+	 * Measurement_Event_Builder constructor.
+	 *
+	 * @param array $config The measurement event configuration.
+	 */
+	public function __construct( array $config ) {
+		$this->configuration = $config;
+	}
+
+	/**
+	 * Gets the name of the plugin
+	 *
+	 * @return string
+	 */
+	public function get_plugin_name() {
+		return $this->configuration['pluginName'];
+	}
+
+	/**
+	 * Gets the name of the event category
+	 *
+	 * @return string
+	 */
+	public function get_category() {
+		return $this->configuration['category'];
+	}
+
+	/**
+	 * Gets the name of the specific event
+	 *
+	 * @return string
+	 */
+	public function get_action() {
+		return $this->configuration['action'];
+	}
+
+	/**
+	 * Gets the CSS selector used to grab element that this event is tied to
+	 *
+	 * @return string
+	 */
+	public function get_selector() {
+		return $this->configuration['selector'];
+	}
+
+	/**
+	 * Gets the inner layer event to bind to in order to track the event
+	 *
+	 * @return string
+	 */
+	public function get_on() {
+		return $this->configuration['on'];
+	}
+
+	/**
+	 * Returns MeasurementEvent object once all params have been set
+	 *
+	 * @return Measurement_Event
+	 */
+	public function build() {
+		return new Measurement_Event( $this );
+	}
+
+}
+
+/**
+ * Represents a single event that ShirshuClass tracks
+ *
+ * @class Measurement_Event
+ */
+class Measurement_Event implements \JsonSerializable {
+
+	/**
+	 * The plugin that this event is associated with
+	 *
+	 * @var string
+	 */
+	private $plugin_name;
+
+	/**
+	 * Event category e.g. ecommerce, engagement
+	 *
+	 * @var string
+	 */
+	private $event_category;
+
+	/**
+	 * Name of specific event
+	 *
+	 * @var string
+	 */
+	private $event_action;
+
+	/**
+	 * CSS selector used to grab element that this event is tied to
+	 *
+	 * @var string
+	 */
+	private $event_selector;
+
+	/**
+	 * Inner layer event to bind to in order to track the event
+	 *
+	 * @var string
+	 */
+	private $event_on;
+
+	/**
+	 * Creates a measurement event builder for the plugin.
+	 *
+	 * @param Array $plugin The measurement event configuration.
+	 *
+	 * @return Measurement_Event_Builder
+	 */
+	public static function create_builder( $plugin ) {
+		return new Measurement_Event_Builder( $plugin );
+	}
+
+	/**
+	 * Measurement_Event constructor.
+	 *
+	 * @param Measurement_Event_Builder $builder The measurement event builder.
+	 */
+	public function __construct( Measurement_Event_Builder $builder ) {
+		$this->plugin_name    = $builder->get_plugin_name();
+		$this->event_category = $builder->get_category();
+		$this->event_action   = $builder->get_action();
+		$this->event_selector = $builder->get_selector();
+		$this->event_on       = $builder->get_on();
```

Comment:
See my above comment, we can directly receive the `array $config` here instead. And then we don't need all these class properties, we could just have a single `$config` class property which we assign the passed value to after validating - this is much simpler than all the properties since we only need the entire `$config` later anyway.

The validation should either use `wp_parse_args` to fill up missing values in the associative array with defaults, or it should check for anything missing and throw an appropriate exception if so. We just need to make sure the `$config` in the class property is valid.